src/damavand/cloud/aws/bucket.py

```diff
@@ -48,11 +51,63 @@ def add_object(self, object: bytes, path: str):
             )
         except ClientError as e:
             logger.error(f"Failed to add object to bucket `{self.name}`: {e}")
+            raise RuntimeError(e)
+
+    @runtime
+    def read(self, path: str) -> bytes:
+        try:
+            buffer = io.BytesIO()
+            self.__s3_client.download_fileobj(
+                Bucket=self.name, Key=path, Fileobj=buffer
+            )
+
+            return buffer.getvalue()
+        except ClientError as e:
+            logger.error(f"Failed to read object at `{path}`: {e}")
+            raise RuntimeError(e)
+
+    @runtime
+    def delete(self, path: str):
+        try:
+            self.__s3_client.delete_object(Bucket=self.name, Key=path)
+        except ClientError as e:
+            logger.error(f"Failed to delete object at `{path}`: {e}")
+            raise RuntimeError(e)
+
+    @runtime
+    def list(self) -> Iterable[str]:
+        """
+        Return an iterable of object keys in the bucket.
+
+        __ATTENTION__: This method is expensive for large buckets as it request multiple times to fetch all objects.
+        """
+        try:
+            paginator = self.__s3_client.get_paginator("list_objects_v2")
+            for page in paginator.paginate(Bucket=self.name):
+                for obj in page.get("Contents", []):
+                    yield obj["Key"]
+        except ClientError as e:
+            logger.error(f"Failed to list objects in storage `{self.name}`: {e}")
+            raise RuntimeError(e)
+
+    @runtime
+    def exist(self, path: str) -> bool:
+        """Check if an object exists in the bucket."""
```

Comment:
What if the object has a delete marker in a versioned bucket?

Reply:
@shahinism you are right on this! My motivation was to not request the object's body inorder to save unnecessary data transfer (cost optimize it). Do you have any suggestion?